tscv/_split.py

```diff
@@ -161,14 +161,27 @@ def __complement_masks(self, masks):
                     complement[np.arange(begin, end)] = False
             yield complement
 
+    def __insert_seq_after(self, arr, n):
+        max_value = np.max(arr)
+        return np.append(arr, np.arange(max_value, max_value+n+1))
+
+    def __insert_seq_before(self, arr, n):
+        min_value = np.min(arr)
+        return np.insert(arr, 0, np.arange(min_value-n, min_value))
+
     def __complement_indices(self, indices, n_samples):
         before, after = self.gap_before, self.gap_after
         for index in indices:
-            complement = np.arange(n_samples)
-            for i in index:
-                begin = max(i - before, 0)
-                end = min(i + after + 1, n_samples)
-                complement = np.setdiff1d(complement, np.arange(begin, end))
+            complement = np.setdiff1d(
```

Comment:
`index` might not be a contiguous set.

Reply:
Hi, sorry, can you provide an example where index is not contiguous?

Reply:
For example, the test set can be `[5, 7, 9]` or `[4, 6, 8]`, which are not contiguous. This kind of test sets is uncommon, but the interface does not forbid the user to do so. That's why we need to handle this corner case (which is actually a more general case).

Reply:
I'm sorry, still can't look a situation like this. in your example if we have a set like `[5, 7, 9]`, then we get the `min - before `and the `max + gap_after etc`, but then, in order to get the complement, we create a range (contiguous) between begin and end, so....

Reply:
All right. Let's consider the following case.
```python
data = [1, 2, 3, 4, 5, 6, 7, 8, 9]
test = [1, 2, 8, 9]
gap_before = 1
gap_after = 1
```

In this case, the training set should be `[4, 5, 6]`, not empty.

Reply:
All right. Let's consider the following case.
```python
data = [1, 2, 3, 4, 5, 6, 7, 8, 9]
test = [1, 2, 8, 9]
gap_before = 1
gap_after = 1
```

In this case, the training set should be `[4, 5, 6]`, not empty.

Reply:
Ok I got it, you're right, this case is not correctly handled with my code, but is more a general case, as you said before. I'm not sure that these kind of splits can be generate by the actual cv methods.

Reply:
What do you mean by this sentence?
> I'm not sure that these kind of splits can be generate by the actual cv methods.

If you tweak your modification a little bit, you can handle the more general case without sacrificing too much the efficiency.

Reply:
I mean that, according to the implementation of `_iter_test_indices` methods, there shouldn't be cases of test folds with uncontiguous elements, but I may be wrong.
Btw I don't think I can find a tweak asap, and I don't have much time right now. Feel free to leave the PR open or delete if as you wish :)

Reply:
> according to the implementation of `_iter_test_indices` methods, there shouldn't be cases of test folds with uncontiguous elements, but I may be wrong.

Indeed, all existing subclasses in this package give contiguous test sets. However, the base class `GapCrossValidator` does not impose this constraint. Users can subclass it and produce uncontiguous test sets:
```python
class Foo(tscv.GapCrossValidator):
    def _iter_test_indices(self, X=None, y=None, groups=None):
        n = len(X)
        yield list(range(0, n, 2))
        yield list(range(1, n, 2))
        
    def get_n_splits(self, X=None, y=None, groups=None):
        return 2

for train_index, test_index in Foo().split(range(10)):
    print("TRAIN:", train_index, "TEST:", test_index)
```

```
TRAIN: [1 3 5 7 9] TEST: [0, 2, 4, 6, 8]
TRAIN: [0 2 4 6 8] TEST: [1, 3, 5, 7, 9]
```

Reply:
If you don't have time, I would suggest making a small improvement by using branching. If the test set is contiguous, then use your code. Otherwise, use the original code. You can test the contiguity with `len(index) == index[-1] - index[0] + 1`.